### DISCUSSION TITLE
Access to Astro.locals inside getStaticPaths()

### DISCUSSION THREAD

**Discussion by koronowicz** [withastro/roadmap#605] [Proposal]
### Body

# Summary

I created a data model for my website inside middleware and access it in .astro files with Astro.locals. Sadly I cannot access Astro.locals inside getStaticPaths(). Is there any specific reason why it's not possible?

# Goals

Access to Astro.locals inside getStaticPaths()

**Comment by koronowicz** [comment]
I found workaround using custom Endpoint to fetch and process data, but I don't think it's not an ideal solution. Endpoint should be hidden and I need to care about fetch URL instead just referring to a method from Astro.locals.

**Comment by ematipico** [comment]
@koronowicz I'd suggest expanding your proposal:
- being more descriptive about your use case and why you need it
- create some pseudo code to show how you'd use the feature you're proposing

For example, I wouldn't be able to help you right now because I can't understand the propsal.

**Reply by koronowicz** [reply]
Sure, 

my use case is product website, I fetch data from multiple API's, after fetching I do some operations on it. I created `middleware.js` where I fetch all my data and return `Astro.locals.website_data` object with data for all my pages. In that case I don't need to call any additional `fetch` in other parts of project + no additional data modification.

```
export function onRequest ({ locals, request }, next) {

  locals.website_data = await fetch('https://example.com/data.json').then(r => r.json())
  // some modifications
  return next()
};
```

Everything worked good until I needed to create `/blog/[slug].astro` pages. There is no access to `Astro.locals` inside  `getStaticPaths()`.

My workaround was to create an `Endpoint` that return the same data as my `Middleware` and I fetch it inside `getStaticPaths()`. It's working well, but now I have a unnecessary endpoint in my website. For Netlify build I needed to separate into two Astro websites: API Endpoint and Real Website that fetch from API Endpoint.

**Reply by ematipico** [reply]
And would you like to access to `Astro.locals`? Have you envisioned that?

**Reply by koronowicz** [reply]
It would simplify our use case. We have 5 different API endpoints to fetch data, some of those API's are slower. At the beginning in dev mode I couldn't do anything, `node-fetch-cache` fix that problem. What I have envisioned is from my experience with Nuxt. It will be good, to have a data store that could be fetch once for build and dev. Then have access to that data in Pages and Components, also inside getStaticPaths().

**Comment by matthewp** [comment]
Astro.locals is scoped to a request. `getStaticPaths` is not called within a request; it's called to figure out pages to build.

**Reply by koronowicz** [reply]
Yes, do you see any workaround to that use case?

**Reply by ematipico** [reply]
We wouldn't be able to fix your use case, I think. As you can understand from the function's name, `getStaticPaths` is meant for _static_ pages.

Also, the middleware lifecycle is tight to the request of a page, which means that the middleware is called **after** Astro collects the page. `getStaticPaths` is essentially called **before** the middleware.

**Reply by koronowicz** [reply]
Thanks for quick answer. We will move that logic to backend service. Still very happy with using Astro ❤️
